nestkernel/secondary_event.h

```diff
@@ -0,0 +1,464 @@
+    left_to_copy -= sizeof( unsigned int );
+  }
+
+  pos += num_uints;
+}
+
+/**
+ * This template function reads data of type T from a given position of a
+ * std::vector< unsigned int >. The function is used to read SecondaryEvents
+ * data from the NEST communication buffer. The pos iterator is advanced
+ * during execution. For a discussion on the functionality of this function see
+ * github issue #181 and pull request #184.
+ */
+template < typename T >
+void
+read_from_comm_buffer( T& d, std::vector< unsigned int >::iterator& pos )
+{
+  // there is no aliasing problem here, since cast to char* invalidate strict
+  // aliasing assumptions
+  char* const c = reinterpret_cast< char* >( &d );
+
+  const size_t num_uints = number_of_uints_covered< T >();
+  size_t left_to_copy = sizeof( T );
+
+  for ( size_t i = 0; i < num_uints; i++ )
+  {
+    memcpy( c + i * sizeof( unsigned int ), &( *( pos + i ) ), std::min( left_to_copy, sizeof( unsigned int ) ) );
+    left_to_copy -= sizeof( unsigned int );
+  }
+
+  pos += num_uints;
+}
+
+/**
+ * Template class for the storage and communication of a std::vector of type
+ * DataType. The class provides the functionality to communicate homogeneous
+ * data of type DataType. The second template type Subclass (which should be
+ * chosen as the derived class itself) is used to distinguish derived classes
+ * with the same DataType. This is required because of the included static
+ * variables in the base class (as otherwise all derived classes with the same
+ * DataType would share the same static variables).
+ *
+ * Technically the DataSecondaryEvent only contains iterators pointing to
+ * the memory location of the std::vector< DataType >.
+ *
+ * Conceptually, there is a one-to-one mapping between a SecondaryEvent
+ * and a SecondaryConnectorModel. The synindex of this particular
+ * SecondaryConnectorModel is stored as first element in the static vector
+ * supported_syn_ids_ on model registration. There are however reasons (e.g.
+ * the usage of CopyModel or the creation of the labeled synapse model
+ * duplicates for pyNN) which make it necessary to register several
+ * SecondaryConnectorModels with one SecondaryEvent. Therefore the synindices
+ * of all these models are added to supported_syn_ids_. The
+ * supports_syn_id()-function allows testing if a particular synid is mapped
+ * with the SecondaryEvent in question.
+ */
+template < typename DataType, typename Subclass >
+class DataSecondaryEvent : public SecondaryEvent
+{
+private:
+  // we chose std::vector over std::set because we expect this to be short
+  static std::vector< synindex > pristine_supported_syn_ids_;
+  static std::vector< synindex > supported_syn_ids_;
+  static size_t coeff_length_; // length of coeffarray
+
+  union CoeffarrayBegin
+  {
+    std::vector< unsigned int >::iterator as_uint;
+    typename std::vector< DataType >::iterator as_d;
```

Comment:
Maybe
```suggestion
    typename std::vector< DataType >::iterator as_DataType;
```
Same below.